### ISSUE TITLE
Dormant chunk cache causes entities loosing data

### ISSUE THREAD

**Issue by leagris** [MinecraftForge/MinecraftForge#3897]
Happens in SMP with `level-type=RTG`

Issue has been previously reported to https://github.com/Team-RTG/Realistic-Terrain-Generation/issues/1168, but they deny their RTG mod is actively involved into this dormant chunk cache misbehavior with entities.

[Team-RTG says there is no RTG code involved in this issue Realistic-Terrain-Generation/Issue 1168](https://github.com/Team-RTG/Realistic-Terrain-Generation/issues/1168#issuecomment-301461850 ). So I guess there is some wierd condition triggered into forge with different level type here.
> no RTG code is executed in any way during the chunk data saving and loading processes.

##### Affected Mod Versions:

- Forge version: either `13.20.0.2282` or `13.20.0.2296` can reproduce bug
- RTG version: `1.11.2-5.1.0.0-alpha 1` (Alone, without any other mod)

##### Issue Description:

When in SMP Forge Server `config/forgeChunkLoading.cfg` is set with `I:dormantChunkCacheSize` > `0` and `level-type=RTG`

Returning to already generated, but later unloaded chunks; causes Living Entities to bug:

- Some entities (random) becomes immobile (no AI).
- Bugged entities can not be interacted with, (no push, no damage, no Villager deal on right-click).
- Server log messages while this happens:

```
[Server thread/WARN]: Fetching addPacket for removed entity
```

As many times as you will find bugged entities.

##### Steps to reproduce the issue: (100% reproducible.)

1. Create an SMP Server with: either Minecraft Forge `13.20.0.2282` or `13.20.0.2296`. and 1. Install RTG `1.11.2-5.1.0.0-alpha 1`
1. In `server.properties`, set:  
    ````
    level-type=RTG
    level-seed=7007098431119981107
    ````
1.  in `config/forgeChunkLoading.cfg`, set:
    ```
    I:dormantChunkCacheSize=256
    ```  
1. Start server and connect to world with client.
1. Reach near-by Village1 at x=250, y=72, z=217.
1. Visit distant Village2 at x=2378, y=95, z=293 or go far enough so chunks containing Village1 gets unloaded by server.
1. Return to Village1 at x=250, y=72, z=217. See some villagers or other near-by animals bugged as described. See server console warning message about Fetching addPacket for removed entity.

##### Extra Information:

1. Bug only happens in SMP with `level-type=RTG` and a non-zero `dormantChunkCacheSize`.
1. Bugged entities are discarded on server restart.
1. Disconnecting/reconnecting client; retrieves same bugged entities unless server has been restarted. (see 2.).

**Comment by mezz** [comment]
My gut reaction is the same as yours, this seems like an RTG bug.
Checking discord, srs-bsns acted very obnoxiously.
I will try to look into this more.

**Comment by mezz** [comment]
This is tough to debug, here's what I've found so far:

- When a dormant chunk with entities is loaded, `chunk#onChunkLoad` is called which loads the entities. 

- During that, `WorldServer#canAddEntity` is called and there is a check to see if the entity already exists:
  `if (this.entitiesByUuid.containsKey(uuid))`

-  For the dormant chunks, the entities are still in `entitiesByUuid` and so they are counted as existing for some reason. I believe this is the bug, they should have been removed? In any case, that check is triggered and the server kills the entity as a failsafe measure.

- Normally killing an entity notifies the client, but there are no entity trackers set up for these entities so the client is never notified and ends up with motionless entities that are actually dead on the server.

I was able to reproduce this with Forge alone, so it is a Forge bug.

**Comment by srs-bsns** [comment]
>Checking discord, srs-bsns acted very obnoxiously.

🙄

**Comment by mezz** [comment]
@srs-bsns Please do what you can to point people in the right direction, try to be constructive and not dismissive. A lot of players are afraid of reporting bugs or just avoid it because of the negative way developers respond to them.

**Comment by MoeBoy76** [comment]
This is the kind of attitude that keeps mod devs from asking questions on how things should be done and why they never send people to Forge, a simple "ah, it *is* a Forge bug, sorry srs_bsns" would make a big difference.

This appeared to be an RTG issue but RTG doesn't do anything to loading or unloading, only changes terrain when it generates initially, which means it had to be a Forge bug as srs_bsns stated in RTG's Discord multiple times.

**Comment by mezz** [comment]
I just spent two and a half hours digging in and trying to match Leagris's effort in order to find the cause of this issue and discover that it was not from RTG, I think you have the wrong idea.
I see now that it looks like I was calling srs-bsns names because I assumed it was their bug. That's not the case and I should have explained earlier. I'm sorry.
If you have any questions you should feel free to ask and when I'm around I will do my best to help out.

**Comment by leagris** [comment]
Thank you for your efforts digging into this @mezz. I take my share of bad day misunderstandings.

Nobody has to look bad here, and surely not @srs-bsns who took time to answer my questions into Team-RTG Discord. Note that I was unable to understand his reason I was being told it is not RTG at fault.

I have a limited knowledge and understanding of mods and Forge internal workings. Sometimes stuffs behaves like it is a mod at fault on the surface, even after spending hours repeating tests to match one specific mod in our in-dev mod-pack.

I mean I am not qualified to sort such issue out independently.
A bit of guidance and initiative to redirect the issue to Forge by one of Team-RTG would have eased it.

Lets focus on the issue here as a final word.

My apologies for making someone look bad while he tried to help.

**Comment by bs2609** [comment]
@mezz, I looked into this earlier and think I found a potential issue, will be making a PR on it shortly.

**Comment by bs2609** [comment]
Here's further analysis of mezz's original findings:

* When a chunk is unloaded `chunk.onChunkUnload()` is called, which calls `world.unloadEntities()`
* This adds the entities to the `unloadedEntityList`.
* During `world.updateEntities()`, this list is iterated over, and `world.onEntityRemoved()` is called for each entity. After the list has been processed, it is cleared.
* `WorldServer#onEntityRemoved()` removes entities from the ID and UUID maps.

If `WorldServer.canAddEntity()` detects a duplicate entity in the UUID map, it checks if it's in the unloaded entity list.
* If it is, it removes it from that list and calls `World#removeEntityDangerously()` on it. It then returns true, signalling the entity can be added. The entity is then added to the world, but it is dead, as `removeEntityDangerously()` also kills it.
* If it is not, a warning about the already existing entity is logged, and `canAddEntity()` promptly returns false. The entity is not removed, and also not re-added.

That would explain much of the strange behaviour seen, but...

Any chunk being loaded calls `onChunkLoad()`, and dormant chunks won't be fetched unless the actual chunk is not loaded, which suggests `onChunkUnload()` would have been called on that chunk previously. In the normal course of events, that would remove entities from the UUID map, and no duplicates would occur.

Outside of what normally should happen, it is possible for entities added to be in the world's UUID map already. If the world in question has not yet ticked it's entities pending the chunk's first removal, `onEntityRemoved()` would not have been called, and so the entity would still be seen there. If that was the case, the entity would also still be in the list of entities pending removal.

This would be a reason for `canAddEntity()` checking `unloadedEntityList.contains(entity)` in the event of an existing UUID being found. The resulting deadness of the entity might be simply an oversight.

With all of that said, ~~there doesn't seem to be anything specific to dormant chunks here~~. It appears to be possible to trigger something like this issue by, in the interval between the chunk provider being ticked (during `world.tick()`, processes chunk unloads), and the world's entities being ticked (during `world.updateEntities()`, processes entity unloads), loading a chunk that has just been unloaded.

~~I'm not sure yet whether there is anything about dormant chunks that makes this more likely.~~

**Comment by bs2609** [comment]
Followup: dormant chunks are exclusively affected by this because they don't discard and recreate their entities.

Vanilla isn't affected, any new entity created will not appear as a member of `unloadedEntityList` and so that code path will not be taken.

Instead, there will be a log entry about "Keeping entity that already exists", then on the world entity tick, the old entity will be removed from the world. As the chunk it should be a part of is now loaded, the removal process will attempt to remove it from the chunks entity lists, but this will likely fail, as the 'old' entity does not appear to exist there. The world entity will most possibly be GC'd at some later time and the chunk entity will be an orphan until that chunk is reloaded.

Entities in a dormant chunk will pass the `unloadedEntityList` check and ends up added to the world in a dead state, causing the issues described.

So not great either way, really.

**Comment by bs2609** [comment]
It's the fact that with dormant chunks, the 'old' and 'new' entities are both the same object, that is responsible for the buggy behaviour here.

**Comment by bs2609** [comment]
This should be fixed now by #3940/#4162.
